File: csa2.js
//compile me with : java -jar node_modules\google-closure-compiler\compiler.jar --language_in=ECMASCRIPT6 --language_out=ES5 --js_output_file=out.js csa.js

var split = require('split');
var isTimetableLoaded = false;
var timeTable = [];
const MAX_STATIONS = 20;
const TRANSFER_TIME_THRESHOLD = 900; //15 minutes

  
//first we get the connextions from stdin, then an empty line, then one request at a time.
function processLine (line) { 
  if(line == ""){
    if(isTimetableLoaded){ //empty line after timetable is loaded is terminating signal.
      process.exit();
    }
    isTimetableLoaded = true;
    return;
  }
  if(!isTimetableLoaded){
    processConnection(line)
  }else{
    processRequest(line)
  }
}

function processConnection(line){
  var tokens = line.split(" ");
  timeTable.push({ 
    departureStation : parseInt(tokens[0]),
    arrivalStation : parseInt(tokens[1]),
    departureTimestamp : parseInt(tokens[2]),
    arrivalTimestamp : parseInt(tokens[3])
  });
}

function processRequest(line){
  var tokens = line.split(" ");
  var request = { 
    departureStation : parseInt(tokens[0]),
    arrivalStation : parseInt(tokens[1]),
    departureTimestamp : parseInt(tokens[2])
  };


  compute(request);

}

function compute(request){
  var earliestArrival = [];
  var earliestArrivalMinConnections = []; //intended to store a list of {departureStation, arrivalTimestamp, ConnectionCount, refersToTimetableIndex}

  //init inConnection and earliestArrival
  for(var i = 0; i < MAX_STATIONS ; i++){     
    earliestArrival[i] = Infinity;    
    earliestArrivalMinConnections[i] = [];
  }

  earliestArrival[request.departureStation] = request.departureTimestamp;
  earliestArrivalMinConnections[request.departureStation].push({departureStation:NaN, 
                                                                departureTimestamp:NaN,
                                                                arrivalTimestamp: request.departureTimestamp, 
                                                                connectionCount: -1, 
                                                                refersToTimetableIndex: NaN,
                                                                inConnection: null,
                                                                minimumTransferTime : Infinity
                                                              });

  //test if exceding MAX_STATIONS
  if(request.departureStation <= MAX_STATIONS && request.arrivalStation <= MAX_STATIONS){
    mainLoop(request, earliestArrivalMinConnections, earliestArrival);
  }

  //display the results
  if(hasSolutionOrReport(request, earliestArrivalMinConnections)){
    printResultEarliest(request, earliestArrivalMinConnections);
    printResultLeastConnections(request,earliestArrivalMinConnections);
    console.log("");//flush answer
  } 


}

function mainLoop(request, earliestArrivalMinConnections, earliestArrival){
  //loop with no optim what so ever.
  //we will store every possibility, options will be eliminated only after trying each timetable item.
  timeTable.forEach(function(connection, indexOnTimetable){
    //one can eliminate values if it departs too early
    //TODO : consider looking up earliestArrival based on data already present in earliestArrivalMinConnections. for space complexity optim at cost of time complexity
    if(connection.departureTimestamp >= earliestArrival[connection.departureStation]){//this connection can be used
      //update earliest arrival at station
      earliestArrival[connection.arrivalStation] = Math.min(connection.arrivalTimestamp, earliestArrival[connection.arrivalStation] ); 
      
      //for each value of earliestArrivalMinConnections[connection.departureStation], 
      earliestArrivalMinConnections[connection.departureStation].forEach(function(value, index){//TODO, what out for index ovverriden. might lead to trouble !
        //see if value.arrivalTimestamp < connection.departureTimestamp.
        //if it is the case, then, look up for the number of connection, 
        if(value.arrivalTimestamp < connection.departureTimestamp){
          //and append a new value to the array earliestArrivalMinConnections[connection.arrivalStation], incrementing the connectionCount
          //TODO : consider not adding if one already arrived at this station earlier with the same connection count, reduces computation and space cost of computeRouteLeastConnections function
          //TODO : consider making minimum transfer time station specific. If so, store a boolean and compare each time with the corresponding threshold
          earliestArrivalMinConnections[connection.arrivalStation].push({departureStation:connection.departureStation, 
                                                                        departureTimestamp:value.departureTimestamp ? value.departureTimestamp : connection.departureTimestamp,//TODO This repeats the information on every node. Consider adding it only to the node departing from origin
                                                                        arrivalTimestamp: connection.arrivalTimestamp, 
                                                                        connectionCount: value.connectionCount+1, 
                                                                        refersToTimetableIndex: indexOnTimetable,
                                                                        inConnection: value,
                                                                        minimumTransferTime: ( value.departureStation ? Math.min(value.minimumTransferTime, connection.departureTimestamp - value.arrivalTimestamp) : value.minimumTransferTime )
                                                                        //only consider it to be a transfer if previsous step is not the initial arrival
                                                                      });

        }
      });
    }
  });  
}


function computeRouteLeastConnections(lastStep, earliestArrivalMinConnections, route){
  //if lastStep is null, we have to base of the request
  //for each possible step at this station, look if we are done
  var possibilities = earliestArrivalMinConnections[lastStep.departureStation];
  
  var selectedPossibility = lastStep.inConnection;
  
  if(!selectedPossibility.refersToTimetableIndex){//found our starting point
    return route;
  }else{
    route.push(timeTable[selectedPossibility.refersToTimetableIndex]);
  }

  return computeRouteLeastConnections(selectedPossibility, earliestArrivalMinConnections, route);
}

function hasSolutionOrReport(request, earliestArrivalMinConnections){
   if(earliestArrivalMinConnections[request.arrivalStation].length == 0){
    console.log("NO_SOLUTION");
    process.stderr.write("NO_SOLUTION\n");
    return false;
  }else{
    return true;
  }
}


function orderStepsBy_EarliestArrival_LatestDeparture(a,b){
  if (a.arrivalTimestamp < b.arrivalTimestamp){
    return -1;
  }else if (a.arrivalTimestamp > b.arrivalTimestamp){
    return 1;
  }else{
    if (a.departureTimestamp > b.departureTimestamp){
      return -1;
    }else if (a.departureTimestamp < b.departureTimestamp){
      return 1;
    }
  }
  return 0;
}

function orderStepsBy_ConnectionCount_EarliestArrival_LatestDeparture(a,b){
  if (a.connectionCount < b.connectionCount){
    return -1;
  }else if (a.connectionCount > b.connectionCount){
    return 1;
  }else{
    if (a.arrivalTimestamp < b.arrivalTimestamp){
      return -1;
    }else if (a.arrivalTimestamp > b.arrivalTimestamp){
      return 1;
    }else{
      if (a.departureTimestamp > b.departureTimestamp){
        return -1;
      }else if (a.departureTimestamp < b.departureTimestamp){
        return 1;
      }
    }
  }
  return 0;
}
  
function printResultEarliest(request, earliestArrivalMinConnections){

  var route = [];

  //find value for which arrivalTimestamp is minumum to get where to start.
  var possibilities = earliestArrivalMinConnections[request.arrivalStation];

  possibilities = possibilities.sort(orderStepsBy_EarliestArrival_LatestDeparture);

  var selectedPossibility = possibilities[0];

  var routeHasShortTransferTime = selectedPossibility.minimumTransferTime < TRANSFER_TIME_THRESHOLD;

  //While we didn't reach the destination, follow the node using the property inConnection
  while(selectedPossibility.inConnection != null){
    route.push(timeTable[selectedPossibility.refersToTimetableIndex]);
    selectedPossibility = selectedPossibility.inConnection;
  }

  route.reverse().forEach(function(connection){
    process.stderr.write("EARLIEST_ARRIVAL : " + connection.departureStation+" "+connection.arrivalStation+" "+connection.departureTimestamp+" "+connection.arrivalTimestamp +"\n");
    //solutionType 1 indicates fastest route
    console.log("EARLIEST_ARRIVAL "+connection.departureStation+" "+connection.arrivalStation+" "+connection.departureTimestamp+" "+connection.arrivalTimestamp );
  });


  if(routeHasShortTransferTime){//compute an alternative route without short transfer.  
    route = [];
    selectedPossibility = null;
    for(var i = 0; i < possibilities.length; i++){
      if(possibilities[i].minimumTransferTime >= TRANSFER_TIME_THRESHOLD){
        selectedPossibility = possibilities[i];
        break;
      }
    } 

    if(selectedPossibility){ //It is possible there is no way to get all the transfers longer than the threshold
      //While we didn't reach the destination, follow the node using the property inConnection
      while(selectedPossibility.inConnection != null){
        route.push(timeTable[selectedPossibility.refersToTimetableIndex]);
        selectedPossibility = selectedPossibility.inConnection;
      }

      route.reverse().forEach(function(connection){
        process.stderr.write("EARLIEST_ARRIVAL_WITH_EASY_TRANSFERS : " + connection.departureStation+" "+connection.arrivalStation+" "+connection.departureTimestamp+" "+connection.arrivalTimestamp +"\n");
        //solutionType 1 indicates fastest route
        console.log("EARLIEST_ARRIVAL_WITH_EASY_TRANSFERS "+connection.departureStation+" "+connection.arrivalStation+" "+connection.departureTimestamp+" "+connection.arrivalTimestamp );
      });
    }
  }   
}


function printResultLeastConnections(request, earliestArrivalMinConnections){

  //for least ammount of connections, one has to start at the arrival, and follow the links towards the start.
  //TODO if several choices, criteria should help choosing. 
  route = [];

  //find value for which connectionCount is minumum to get where to start.
  var possibilities = earliestArrivalMinConnections[request.arrivalStation];

  possibilities = possibilities.sort(orderStepsBy_ConnectionCount_EarliestArrival_LatestDeparture);

  var selectedPossibility = possibilities[0];

  var routeHasShortTransferTime = selectedPossibility.minimumTransferTime < TRANSFER_TIME_THRESHOLD;

  route.push(timeTable[selectedPossibility.refersToTimetableIndex]);
  route = computeRouteLeastConnections(selectedPossibility, earliestArrivalMinConnections, route);

  route.reverse().forEach(function(connection){
    process.stderr.write("LEAST_CONNECTIONS " + connection.departureStation+" "+connection.arrivalStation+" "+connection.departureTimestamp+" "+connection.arrivalTimestamp +"\n");
    //solution type 2 indicates least connection number route
    console.log("LEAST_CONNECTIONS "+connection.departureStation+" "+connection.arrivalStation+" "+connection.departureTimestamp+" "+connection.arrivalTimestamp );
  });  

  if(routeHasShortTransferTime){//compute an alternative route without short transfer.

    route = [];
    selectedPossibility = null;

    for(var i = 0; i < possibilities.length; i++){
      if(possibilities[i].minimumTransferTime >= TRANSFER_TIME_THRESHOLD){
        selectedPossibility = possibilities[i];
        break;
      }
    }

    if(selectedPossibility){ //It is possible there is no way to get all the transfers longer than the threshold

      route.push(timeTable[selectedPossibility.refersToTimetableIndex]);
      route = computeRouteLeastConnections(selectedPossibility, earliestArrivalMinConnections, route);

      route.reverse().forEach(function(connection){
        process.stderr.write("LEAST_CONNECTIONS_WITH_EASY_TRANSFERS " + connection.departureStation+" "+connection.arrivalStation+" "+connection.departureTimestamp+" "+connection.arrivalTimestamp +"\n");
        //solution type 2 indicates least connection number route
        console.log("LEAST_CONNECTIONS_WITH_EASY_TRANSFERS "+connection.departureStation+" "+connection.arrivalStation+" "+connection.departureTimestamp+" "+connection.arrivalTimestamp );
      });  

    }




  } 

}


//start looking what comes in from stdin
process.stdin.pipe(split()).on('data', processLine);

File: csa2.rb
#!/usr/bin/env ruby

require 'json'
require "yaml"

MAX_STATIONS = 20
INF = 1.0/0

class Connection
  attr_reader :departure_station, :arrival_station,
              :departure_timestamp, :arrival_timestamp

  def initialize(line)
    tokens = line.split(" ")

    @departure_station    = tokens[0].to_i
    @arrival_station      = tokens[1].to_i
    @departure_timestamp  = tokens[2].to_i
    @arrival_timestamp    = tokens[3].to_i
  end
end

class Timetable
  attr_reader :connections

  # reads all the connections from stdin
  def initialize
    @connections = []
    line = STDIN.gets.strip

    while !line.empty?
      @connections << Connection.new(line)
      line = STDIN.gets.strip
    end
  end
end

class Route_Step
  include Comparable
  attr_reader :trip_departure_timestamp, :arrival_timestamp,
              :connection_count, :index_in_timetable, :previous_step, :comparator_earliestarrival

  def initialize(trip_departure_timestamp, arrival_timestamp, connection_count, index_in_timetable, previous_step)
    @trip_departure_timestamp = trip_departure_timestamp
    @arrival_timestamp = arrival_timestamp
    @connection_count = connection_count
    @index_in_timetable = index_in_timetable
    @previous_step = previous_step
  end
end

comparator_earliestarrival = lambda { |a,b|

    if a.arrival_timestamp < b.arrival_timestamp
      return -1
    elsif (a.arrivalTimestamp > b.arrivalTimestamp)
      return 1
    end
    0
  }


class CSA
  attr_reader :timetable, :possible_routes, :earliest_arrival

  def initialize
    @timetable = Timetable.new
  end

  def main_loop()
    timetable.connections.each_with_index do |c, i|
      possible_routes[c.departure_station].each_with_index do |previous, j|#for each possible previous step
        if c.departure_timestamp >=  previous.arrival_timestamp
          if c.arrival_timestamp < earliest_arrival[c.arrival_station]
            earliest_arrival[c.arrival_station] = c.arrival_timestamp
          end
          possible_routes[c.arrival_station] << Route_Step.new(previous.trip_departure_timestamp ? previous.trip_departure_timestamp : c.departure_timestamp, #if not set, it is the iteration of the outter loop
                                                              c.arrival_timestamp,
                                                              previous.connection_count+1,
                                                              i,
                                                              previous
                                                            )
        end
      end
    end

  end

  def print_result(arrival_station)
    if possible_routes[arrival_station].length == 0
      puts "NO_SOLUTION"
    else
      route = []
      # We have to rebuild the route from the arrival station
      possible_routes[arrival_station].sort_by {|e| e.arrival_timestamp}
      current_route_step = possible_routes[arrival_station][0]

      while current_route_step.connection_count != -1
        route << timetable.connections[current_route_step.index_in_timetable]
        current_route_step = current_route_step.previous_step
      end


      # And now print it out in the right direction
      route.reverse.each do |c|
        STDERR.puts "#{c.departure_station} #{c.arrival_station} #{c.departure_timestamp} #{c.arrival_timestamp}"
        puts "#{c.departure_station} #{c.arrival_station} #{c.departure_timestamp} #{c.arrival_timestamp}"
      end
    end
    puts ""
    STDOUT.flush
  end

  def compute(departure_station, arrival_station, departure_time)
    @possible_routes = Array.new(MAX_STATIONS)
    @earliest_arrival = {}

    MAX_STATIONS.times do |i|
      possible_routes[i] = Array.new()
      earliest_arrival[i] = INF
    end

    earliest_arrival[departure_station] = departure_time;
    possible_routes[departure_station] << Route_Step.new(nil, #not a real departure, just a fake route_step
                                                            departure_time,
                                                            -1,
                                                            nil,
                                                            nil
                                                            )

        
    if departure_station <= MAX_STATIONS && arrival_station <= MAX_STATIONS
      main_loop()
    end

    print_result(arrival_station)
  end
end

def main
  csa = CSA.new

  line = STDIN.gets.strip

  while !line.empty?
    tokens = line.split(" ")
    csa.compute(tokens[0].to_i, tokens[1].to_i, tokens[2].to_i)
    line = STDIN.gets.strip
  end
end

main
